### PR TITLE
DataTracks integration through FFI

### PR DESCRIPTION
**Pull request by stephen-derosa**
# Overview
Integrate DataTracks into the CPP SDK through the FFI

## Details
* setOnAudio/VideoFrameCallback() change signature to use track name
* setOnAudio/VideoFrameCallback() addition signature to use track name
* result.h file for floating up specific errors provided by the FFI
* update token gen script to run for all tokens needed
* Update client-sdk-rust submodule
* data_track_subscription: on pushFrame() called when rust sends new frame, assert(since that breaks cpp<->ffi<->rust agreement of rust handling buffering

## Testing
- src/tests/integration/test_data_track.cpp

## Examples
- [`examples/ping_pong`](https://github.com/livekit-examples/cpp-example-collection/pull/4/changes#diff-9c00330cf3fd9ea6030041f7efae9effe1a2acb068d2c9353ef1a0d0918afedd)

## Blocked By
[Rust SDKs PR](https://github.com/livekit/rust-sdks/pull/905)

Resolves BOT-268

### CODE REVIEW

File: include/livekit/data_track_frame.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2025 LiveKit
```
Comment (xianshijing-lk) [inline include/livekit/data_track_frame.h]: nit, 2026?

File: include/livekit/data_track_info.h [hunk 2]
```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2025 LiveKit
```
Comment (xianshijing-lk) [inline include/livekit/data_track_info.h]: nit, 2026

File: include/livekit/local_data_track.h [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  ~LocalDataTrack() = default;
+
+  LocalDataTrack(const LocalDataTrack &) = delete;
+  LocalDataTrack &operator=(const LocalDataTrack &) = delete;
+
+  /// Metadata about this data track.
+  const DataTrackInfo &info() const noexcept { return info_; }
+
+  /// Raw FFI handle id for internal use.
+  uintptr_t ffi_handle_id() const noexcept { return handle_.get(); }
```
Comment (xianshijing-lk) [inline include/livekit/local_data_track.h]: why this needs to be public ?

File: include/livekit/local_participant.h [hunk 4]
```diff
@@ -135,6 +142,20 @@ class LocalParticipant : public Participant {
...
+   * Data tracks carry arbitrary binary frames and are independent of the
+   * audio/video track hierarchy. The returned LocalDataTrack can push
+   * frames via tryPush() and be unpublished via unpublish().
+   *
+   * @param name  Unique track name visible to other participants.
+   * @return Shared pointer to the published data track.
+   * @throws std::runtime_error on FFI or publish failure.
+   */
+  std::shared_ptr<LocalDataTrack>
+  publishDataTrack(const std::string &name);
```
Comment (xianshijing-lk) [inline include/livekit/local_participant.h]: why we need a publishDataTrack() here ? 

Why can't we re-use the same publishTrack and unpublishTrack ? 
Ideally, users will need to learn somethings specific for data track, it is like, if they know how to use audio track, then they know how to use video track, and data track.

File: .github/workflows/builds.yml [hunk 5]
```diff
@@ -7,6 +7,7 @@ on:
       - src/**
       - include/**
       - examples/**
+      - bridge/**
```
[the remark this answers is not in the document]
Comment (stephen-derosa) [reply .github/workflows/builds.yml]: sounds good -- i will make this a single `git mv` commit at the end so comments are easier to follow!

File: include/livekit/local_data_track.h [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+  /// Whether the track is still published in the room.
+  bool isPublished() const;
+
+  /**
+   * Unpublish this data track from the room.
+   *
+   * After this call, tryPush() will fail and the track cannot be
+   * re-published. The underlying FFI handle is released.
+   */
+  void unpublish();
```
[the remark this answers is not in the document]
Comment (stephen-derosa) [reply include/livekit/local_data_track.h]: fair point, i had thought it was a little cleaner, but it doesnt follow the form of Audio/Video tracks. Updated.

**Review by ladvoc — COMMENTED** [review]
Some initial comments, but I will make another pass. Looking clean so far!

File: include/livekit/data_track_subscription.h [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+ *   while (sub->read(frame)) {
+ *     // process frame.payload
+ *   }
+ */
+class DataTrackSubscription {
+public:
+  struct Options {
+    /// Maximum buffered frames (count). 0 = unbounded.
+    /// When non-zero, behaves as a ring buffer (oldest dropped on overflow).
+    std::size_t capacity{0};
```
[the remark this answers is not in the document]
Comment (stephen-derosa) [reply include/livekit/data_track_subscription.h]: in general i lean towards putting things on the rust side for as you say uniform SDKs. But i dont there should be anything stopping client SDKs from implementing a capacity themselves. Plus in lieu of a rust configurable size this is a great stop gap.

File: include/livekit/data_track_subscription.h [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+private:
+  friend class RemoteDataTrack;
+
+  DataTrackSubscription() = default;
+
+  void init(FfiHandle subscription_handle, const Options &options);
+
+  void onFfiEvent(const proto::FfiEvent &event);
+  void pushFrame(DataFrame &&frame);
```
Comment (ladvoc) [inline include/livekit/data_track_subscription.h]: issue: Data flow is unidirectional (i.e. publisher → subscriber), so there shouldn't be a push frame method for the subscription.

File: src/local_data_track.cpp [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+  msg->set_track_handle(static_cast<uint64_t>(handle_.get()));
+  auto *pf = msg->mutable_frame();
+  pf->set_payload(frame.payload.data(), frame.payload.size());
+  if (frame.user_timestamp.has_value()) {
+    pf->set_user_timestamp(frame.user_timestamp.value());
+  }
+
+  proto::FfiResponse resp = FfiClient::instance().sendRequest(req);
+  const auto &r = resp.local_data_track_try_push();
+  return !r.has_error();
```
Comment (ladvoc) [inline src/local_data_track.cpp]: question: Should we be exposing the error reason to the caller? There are several reasons why pushing a frame could fail (e.g., the track is unpublished, buffer full, etc.). For the ROS bridge, a boolean might be fine if there is logging, but for the core SDK, I lean towards we should report the error reason.

Related question: the FFI interface currently flattens all errors to strings for simplicty, however, on the Rust side there are proper error enums that describe the reason for the error. Should I expose these over FFI rather than using strings? Can an enum describing the error reason map nicely into a C++ exception?

**Review by ladvoc — COMMENTED** [review]
Some additional comments.

File: include/livekit/data_track_subscription.h [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+class DataTrackSubscription {
+public:
+  struct Options {
+    /// Maximum buffered frames. 0 = unbounded.
+    /// When non-zero, behaves as a ring buffer (oldest dropped on overflow).
+    std::size_t capacity{0};
+  };
+
+  virtual ~DataTrackSubscription();
+
```
[the remark this answers is not in the document]
Comment (ladvoc) [reply include/livekit/data_track_subscription.h]: > At least one confusing thing about "subscription" is that this object can also call pushFrame(), so its not purely a subscription

From sync: data flow is unidirectional, so there should be no push method on the subscription.

File: include/livekit/data_track_subscription.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (ladvoc) [reply include/livekit/data_track_subscription.h]: Yes, planning to expose that in the FFI layer today.

File: bridge/src/livekit_bridge.cpp
[the remark this answers is not in the document]
Comment (stephen-derosa) [reply bridge/src/livekit_bridge.cpp]: done

**Review by xianshijing-lk — DISMISSED** [review]

File: .github/workflows/builds.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (stephen-derosa) [reply .github/workflows/builds.yml:10]: no longer relevant with the bridge being deprecated :)

File: src/data_track_subscription.cpp [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+void DataTrackSubscription::pushFrame(DataFrame &&frame) {
+  std::lock_guard<std::mutex> lock(mutex_);
+
+  if (closed_ || eof_) {
+    return;
+  }
+
+  // rust side handles buffering, so we should only really ever have one item
+  if (queue_.size() >= 2) {
+    LK_LOG_ERROR("[DataTrackSubscription] Queue size is greater than 1, this "
```
[the remark this answers is not in the document]
Comment (stephen-derosa) [reply src/data_track_subscription.cpp]: done

File: examples/hello_livekit/hello_livekit.cpp [hunk 12]
```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright 2025 LiveKit
```
[the remark this answers is not in the document]
Comment (stephen-derosa) [reply examples/hello_livekit/hello_livekit.cpp]: done

File: examples/hello_livekit/hello_livekit.cpp [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+  std::signal(SIGINT, handleSignal);
+#ifdef SIGTERM
+  std::signal(SIGTERM, handleSignal);
+#endif
+
+  livekit::initialize(livekit::LogLevel::Info, livekit::LogSink::kConsole);
+
+  // Shared: sender publishes its identity here so the receiver knows who
+  // to subscribe to (no hardcoded identity required).
+  std::atomic<std::string *> sender_identity{nullptr};
```
[the remark this answers is not in the document]
Comment (stephen-derosa) [reply examples/hello_livekit/hello_livekit.cpp]: i broke this into two executables for easier readability/simpler

File: src/data_track_subscription.cpp [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+
+void DataTrackSubscription::pushFrame(DataFrame &&frame) {
+  std::lock_guard<std::mutex> lock(mutex_);
+
+  if (closed_ || eof_) {
+    return;
+  }
+
+  // rust side handles buffering, so we should only really ever have one item
+  if (frame_.has_value()) {
```
[the remark this answers is not in the document]
Comment (stephen-derosa) [reply src/data_track_subscription.cpp]: done since if frame_ had a value, it would break the "rust handle buffers" contract we have.

File: include/livekit/data_track_frame.h [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+  /** Arbitrary binary payload (the frame contents). */
+  std::vector<std::uint8_t> payload;
+
+  /**
+   * Optional application-defined timestamp.
+   *
+   * The proto field is a bare uint64 with no prescribed unit.
+   * By convention the SDK examples use microseconds since the Unix epoch.
+   */
+  std::optional<std::uint64_t> user_timestamp;
```
Comment (xianshijing-lk) [inline include/livekit/data_track_frame.h:46]: nit, suggested adding things like 

DataFrame() = default;
DataFrame(DataFrame&&) = default;
DataFrame& operator=(DataFrame&&) = default;

These constructors will help compiler optimize those move, assignment use cases

File: src/local_data_track.cpp (hunk 9 above)
Comment (xianshijing-lk) [reply src/local_data_track.cpp]: for SDKs used in real-time systems, in general, exceptions are discouraged for "expected" failures like "buffer full" or "track unpublished".

I think the best practice here might be returning an result, which will include the success or error codes/Enums ?

File: include/livekit/data_track_subscription.h [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+ *
+ * Typical usage:
+ *
+ *   auto sub = remoteDataTrack->subscribe();
+ *   DataFrame frame;
+ *   while (sub->read(frame)) {
+ *     // process frame.payload
+ *   }
+ */
+class DataTrackSubscription {
```
Comment (xianshijing-lk) [inline include/livekit/data_track_subscription.h]: maybe I asked this question before, I wonder if should be called data_track_stream ? 
or remote_data_track_stream ? 

so it will be aligned with audio_stream / video_stream

File: src/data_track_subscription.cpp [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+}
+
+void DataTrackSubscription::onFfiEvent(const FfiEvent &event) {
+  if (event.message_case() != FfiEvent::kDataTrackSubscriptionEvent) {
+    return;
+  }
+
+  const auto &dts = event.data_track_subscription_event();
+  if (dts.subscription_handle() !=
+      static_cast<std::uint64_t>(subscription_handle_.get())) {
```
Comment (xianshijing-lk) [inline src/data_track_subscription.cpp]: this worries me a bit, onFfiEvent is a callback. If DataTrackSubscription is moved to a new memory location, the lambda captured this.

take a step back, why do we need to support the move operator ? do we see real use cases from it ?

File: src/data_track_subscription.cpp [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+  }
+
+  // Signal the Rust side that we're ready to receive the next frame.
+  // The Rust SubscriptionTask uses a demand-driven protocol: it won't pull
+  // from the underlying stream until notified via this request.
+  proto::FfiRequest req;
+  auto *msg = req.mutable_data_track_subscription_read();
+  msg->set_subscription_handle(
+      static_cast<uint64_t>(subscription_handle_.get()));
+  FfiClient::instance().sendRequest(req);
```
Comment (xianshijing-lk) [inline src/data_track_subscription.cpp]: I think there are some races there. 
like one thread calls this read(), and pass this section under std::lock_guard<std::mutex> lock(mutex_); and get to this section.

meanwhile, close() is being called, and reset the subscription_handle_

File: src/data_track_subscription.cpp [hunk 19]
```diff
@@ -0,0 +1,182 @@
...
+
+  std::unique_lock<std::mutex> lock(mutex_);
+  cv_.wait(lock, [this] { return frame_.has_value() || eof_ || closed_; });
+
+  if (closed_ || (!frame_.has_value() && eof_)) {
+    return false;
+  }
+
+  out = std::move(frame_.value());
+  frame_.reset();
```
Comment (xianshijing-lk) [inline src/data_track_subscription.cpp:66]: nit, If the DataFrame have move constructor, you don't need this frame_.reset()

File: include/livekit/local_data_track.h [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+   */
+  bool tryPush(const std::vector<std::uint8_t> &payload,
+               std::optional<std::uint64_t> user_timestamp = std::nullopt);
+  /**
+   * Try to push a frame to all subscribers of this track.
+   *
+   * @return true on success, false if the push failed (e.g. back-pressure
+   *         or the track has been unpublished).
+   */
+  bool tryPush(const std::uint8_t *data, std::size_t size,
```
Comment (xianshijing-lk) [inline include/livekit/local_data_track.h]: curiously, do we need all these 3 overloaded methods ? or we should push for one instead?

File: src/local_participant.cpp [hunk 21]
```diff
@@ -286,6 +288,30 @@ LocalParticipant::PublicationMap LocalParticipant::trackPublications() const {
   return out;
 }
 
+std::shared_ptr<LocalDataTrack>
+LocalParticipant::publishDataTrack(const std::string &name) {
```
Comment (xianshijing-lk) [inline src/local_participant.cpp:292]: curiously, what will happen if calling publishDataTrack with the same |name| more than once ?

File: src/local_data_track.cpp [hunk 22]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+bool LocalDataTrack::tryPush(const std::vector<std::uint8_t> &payload,
+                             std::optional<std::uint64_t> user_timestamp) {
+  DataFrame frame;
+  frame.payload = payload;
```
Comment (xianshijing-lk) [inline src/local_data_track.cpp]: this makes me wondering if we should allow developers to pass things like 
bool LocalDataTrack::tryPush(std::vector<std::uint8_t>&& payload,
                             std::optional<std::uint64_t> user_timestamp) { ... }

So that we can use the move constructor for the payload ?

File: src/local_participant.cpp (hunk 21 above)
Comment (stephen-derosa) [reply src/local_participant.cpp:292]: this will throw -- `publishDataTrack()` calls `FfiClient::publishDataTrackAsync` which boils into:
```
  proto::FfiRequest req;
  auto *msg = req.mutable_publish_data_track();
  msg->set_local_participant_handle(local_participant_handle);
  msg->mutable_options()->set_name(track_name);
  msg->set_request_async_id(async_id);

  try {
    proto::FfiResponse resp = sendRequest(req);
    if (!resp.has_publish_data_track()) {
      logAndThrow("FfiResponse missing publish_data_track");
    }
  } catch (...) {
    cancelPendingByAsyncId(async_id);
    throw;
  }
```

this matches the behavior of the `FfiClient::publishTrackAsync`

File: src/local_data_track.cpp (hunk 22 above)
Comment (stephen-derosa) [reply src/local_data_track.cpp]: yeah great point -- much more efficient. I wonder if we even need:
```
bool LocalDataTrack::tryPush(const std::uint8_t *data, std::size_t size,
                             std::optional<std::uint64_t> user_timestamp) {
```
now, might be convenient for some users but it enables stuff to be slower and another step of Q: "hey why is this slow" A: "are you using the right tryPush() function?"

any reason not to keep the single more efficient tryPush()?

File: include/livekit/local_data_track.h (hunk 20 above)
Comment (stephen-derosa) [reply include/livekit/local_data_track.h]: great point, and we have four `tryPush()` calls after adding
```
bool tryPush(std::vector<std::uint8_t> &&payload,
               std::optional<std::uint64_t> user_timestamp = std::nullopt);
```
I think this is too much. My opinion here would be to keep the efficient tryPush() call which uses std::move(), and keep:
```
bool tryPush(const DataFrame &frame);
```

I can forsee `bool tryPush(const DataFrame &frame);` being used to directly forward data from one room to another.

File: src/data_track_subscription.cpp (hunk 19 above)
Comment (stephen-derosa) [reply src/data_track_subscription.cpp:66]: sweet! thank you!

File: include/livekit/data_track_frame.h (hunk 15 above)
Comment (stephen-derosa) [reply include/livekit/data_track_frame.h:46]: ah great point, updated!

File: src/data_track_subscription.cpp (hunk 17 above)
Comment (stephen-derosa) [reply src/data_track_subscription.cpp]: This is a really good catch.

Since the public API provides a shared_ptr<DataTrackSubscription>, i dont think we do need to provide a move constructor. the pointer itself can be std::move(), not the object itself.

Side note:  I wonder how we can setup some tests to try to catch edge cases like this.

File: src/data_track_subscription.cpp (hunk 18 above)
Comment (stephen-derosa) [reply src/data_track_subscription.cpp]: Nice thank you -- with the addition of this request call to the data tracks API i missed putting this in the lock!

File: include/livekit/local_data_track.h (hunk 20 above)
Comment (ladvoc) [reply include/livekit/local_data_track.h]: I agree, we should anticipate data track frames having more metadata fields in the future, and we likely won't want overloads for all of them. Also, small nitpick: in the other SDKs `DataFrame` is called `DataTrackFrame`.

File: src/data_track_subscription.cpp (hunk 18 above)
Comment (stephen-derosa) [reply src/data_track_subscription.cpp]: I also added better logic to close() which:
- gets the lock
- sets closed_ = true
- resets the subscription handle
- sets listener_id_ = 0
- releases the lock
- calls ffi RemoveListener()

File: src/local_data_track.cpp (hunk 9 above)
Comment (stephen-derosa) [reply src/local_data_track.cpp]: thanks for the guidance. I like the enum idea. I will implement something like:
```
Enum ErrorCodes
{
  QUEUE_FULL,
  ...,
  UNKOWN
}
```

I will also adopt this going forward when applicable!

File: include/livekit/data_track_subscription.h (hunk 16 above)
Comment (stephen-derosa) [reply include/livekit/data_track_subscription.h]: i had originally named as data_track_subscription because:
a. data tracks follow a slightly different format than video/audio
b. conceptually it was easier for me to understand.

However, i think your point is especially valid because
a. only data track _publish_ and _unpublish_ are different
b. consistency is important. 

Will update once all other comments are resolved to avoid confusing referencing of MR comments!

File: src/local_data_track.cpp (hunk 9 above)
Comment (ladvoc) [reply src/local_data_track.cpp]: Note: on the Rust side, all fallible operations define error enums which enumerate all possible reasons for the failure. Currently, these are converted to strings when passed over FFI, but we can expose the actual enums if that's helpful.

File: include/livekit/data_track_subscription.h (hunk 16 above)
Comment (ladvoc) [reply include/livekit/data_track_subscription.h]: I would push for keeping it as subscription since there are semantic differences between the two:
- _AudioStream_ and _VideoStream_ for are purely local frame taps. Dropping one removes a local sink, but the SFU subscription is unaffected.
- _DataTrackSubscription_ controls the subscription itself; the first _subscribe()_ call triggers SFU signaling, and dropping the last subscription for a track unsubscribes from the SFU. It's not a passive reader, but rather managing a resource.

File: src/local_data_track.cpp (hunk 9 above)
Comment (stephen-derosa) [reply src/local_data_track.cpp]: i went here with returning a templated `Result` which has the form:
```
template <typename T, typename E> class Result
```

An example is `<std::shared_ptr<DataTrackSubscription>, DataTrackErrorCode>` . I think this is a _little_ complex for the current state but it should scale really well for new error codes! LMK what you think, we can simplify if need be.

Comment (stephen-derosa) [reply src/local_data_track.cpp]: @ladvoc we can definelty support the enums! As of now we are kind of bastardizing them by setting them custom based on some checks. For example stuff like:
```
  if (!handle_.valid()) {
    return Result<void, DataTrackError>::failure(
        DataTrackError{DataTrackErrorCode::INVALID_HANDLE,
                       "LocalDataTrack::tryPush: invalid FFI handle", false});
  }
```

File: src/local_data_track.cpp (hunk 22 above)
Comment (xianshijing-lk) [reply src/local_data_track.cpp]: I would think we will just keep the single more efficient tryPush() so that developers won't pay for the cost without knowing those might involve extra memcopy under the hood. Those performance might be a bottleneck if they are sending high frequency large data.

File: src/data_track_subscription.cpp (hunk 18 above)
Comment (xianshijing-lk) [reply src/data_track_subscription.cpp]: I guess as long as you release the log in the middle of the read function, and close() function reset the subscription_handle_, you will still run into the race here where the invalid subscription_handle_ is being used

Comment (xianshijing-lk) [reply src/data_track_subscription.cpp]: maybe it is better to double check it with the AI ?

File: include/livekit/local_data_track.h (hunk 20 above)
Comment (stephen-derosa) [reply include/livekit/local_data_track.h]: will rename this to DataTrackFrame to align with other SDKs once other comments are resolved for easier tracing.

File: src/local_data_track.cpp (hunk 22 above)
Comment (stephen-derosa) [reply src/local_data_track.cpp]: duplicate thread of here:
https://github.com/livekit/client-sdk-cpp/pull/66#discussion_r3002369500

File: include/livekit/data_track_subscription.h (hunk 16 above)
Comment (stephen-derosa) [reply include/livekit/data_track_subscription.h]: if the trade off is align with cpp or align with other SDKs, aligning with other SDKs is more future proofing.

File: src/data_track_subscription.cpp (hunk 18 above)
Comment (stephen-derosa) [reply src/data_track_subscription.cpp]: i think the race is gated by checks of `if (closed_) return;` before subscription_handle_ is used (except for init).

I did ask Codex GPT 5.4 though and it said:
```
In the current read() path, the handle is read and used while mutex_ is still held in src/data_track_subscription.cpp (line 39). Specifically, read():

- takes mutex_ at src/data_track_subscription.cpp (line 41)
- checks closed_ / eof_
- copies subscription_handle_.get() to a local at src/data_track_subscription.cpp (line 49)
- sends the FFI read request at src/data_track_subscription.cpp (line 58)
```

File: include/livekit/data_frame.h [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+  DataFrame(const DataFrame&) = default;
+  DataFrame(DataFrame&&) noexcept = default;
+  DataFrame& operator=(const DataFrame&) = default;
+  DataFrame& operator=(DataFrame&&) noexcept = default;
+
+  explicit DataFrame(std::vector<std::uint8_t>&& p,
+                     std::optional<std::uint64_t> ts = std::nullopt) noexcept
+      : payload(std::move(p)), user_timestamp(ts) {}
+
+  static DataFrame fromOwnedInfo(const proto::DataTrackFrame &owned);
```
Comment (xianshijing-lk) [inline include/livekit/data_frame.h]: nit, add a comment that this fromOwnedInfo() is used internally only, not public API.

File: include/livekit/data_track_subscription.h (hunk 16 above)
Comment (xianshijing-lk) [reply include/livekit/data_track_subscription.h]: @ladvoc , can you elaborate a bit more ? 
From what I am seeing DataTrackSubscription class does not have subscribe() function.
It is created with auto sub_result = remoteDataTrack->subscribe();
From what I am seeing, developers can subscript to the track multiple times, similar to audioStream / videoStream.

I don't think DataTrackSubscription control the subscription, but the DataTrack does.

**Comment by xianshijing-lk** [comment]
taking a final look

File: src/local_data_track.cpp (hunk 9 above)
Comment (stephen-derosa) [reply src/local_data_track.cpp]: this has been completed.

File: include/livekit/data_track_subscription.h (hunk 16 above)
Comment (stephen-derosa) [reply include/livekit/data_track_subscription.h]: based on this [thread](https://live-kit.slack.com/archives/C09G6SCR620/p1775000773631949?thread_ts=1774901967.943499&cid=C09G6SCR620) i renamed `data_track_subscription` -> `data_track_stream`

File: include/livekit/local_data_track.h (hunk 20 above)
Comment (stephen-derosa) [reply include/livekit/local_data_track.h]: done

File: include/livekit/data_track_subscription.h (hunk 16 above)
Comment (xianshijing-lk) [reply include/livekit/data_track_subscription.h]: Thanks

**Review by xianshijing-lk — APPROVED** [review]
